peak/auto_assembler.py

```diff
@@ -186,23 +191,30 @@ def visit_Attribute(self, node):
                 value = eval(compile(ast.Expression(node.value),
                                      filename="<ast>", mode="eval"),
                              self._locals, self._globals)
-                if getattr(value, "_is_magma_enum", False):
-                    return ast.Call(
-                        ast.Name("assemble", ast.Load()),
-                        [node], []
-                    )
+                if value in self.assemblers:
+                    node_val = eval(compile(ast.Expression(node),
+                                            filename="<ast>", mode="eval"),
+                                    self._locals, self._globals)
+                    bv_type, assembler = self.assemblers[value]
+                    return ast.Num(int(assembler(getattr(bv_type, node.attr))))
             except NameError:
                 pass
         return node
 
 
-def assemble_values_in_func(assembler, peak_fn, _locals, _globals):
+def assemble_values_in_func(assemblers, peak_fn, _locals, _globals):
     func_def = m.ast_utils.get_ast(peak_fn).body[0]
-    func_def = ISABuilderAssembler(_locals, _globals).visit(func_def)
-    # Uncomment to see result of AST rewrite
-    # import astor
-    # print(astor.to_source(func_def))
+    func_def = ISABuilderAssembler(assemblers, _locals, _globals).visit(func_def)
     func_def = ast.fix_missing_locations(func_def)
-    exec(compile(ast.Module([func_def]), filename="<ast>", mode="exec"),
-         _locals, _globals)
+    temp_dir = tempfile.TemporaryDirectory()
+    file_name = os.path.join(temp_dir.name, peak_fn.__name__ + ".py")
+    with open(file_name, "w") as fp:
+        fp.write(astor.to_source(func_def))
+    try:
+        exec(compile(ast.Module([func_def]), filename=file_name, mode="exec"),
+             _globals, _locals)
+    except:
+        tb = traceback.format_exc()
+        print(tb)
```

Comment:
I would prefer this was logged with logging.log rather than printed

Reply:
I partially disagree with that, if anything it should be printed to stderr or use logging.error.

This is special error handling for the case of an exception during the execution of the string/file code (because of the indirection through the file). What this does is: when there's an exception in execing the file, it generates a traceback that points to the file. See the `raise` exception on the next line (it says to view the printed output). So I feel that it's not really a log message, it's either a `logging.error` message or print to `stderr`, since you care, your pick, I'll make the change.

Reply:
(also) the reason that we do this is to get a more meaningful error message, because the default exception raised from the exec isn't as useful, so this is purely in a code path for dumping an error message

Reply:
Logging.error is fine, I just don't want errors / debugging information reported via print.

Reply:
Done

Reply:
I thought about it more, it actually does make sense to use `logging.error`, because if you catch the exception outside this scope, it would make sense to not have the print out, but having it in the error log might be useful for debugging that logic.